lib/AST/Type.cpp

```diff
@@ -4376,32 +4376,33 @@ case TypeKind::Id:
   case TypeKind::BoundGenericClass:
   case TypeKind::BoundGenericEnum:
   case TypeKind::BoundGenericStruct: {
-    auto bound = cast<BoundGenericType>(base);
-    SmallVector<Type, 4> substArgs;
+    const auto bound = cast<BoundGenericType>(base);
     bool anyChanged = false;
-    Type substParentTy;
-    if (auto parentTy = bound->getParent()) {
-      substParentTy = parentTy.transformRec(fn);
-      if (!substParentTy)
+    Type newParentTy;
+    if (const auto parentTy = bound->getParent()) {
+      newParentTy = parentTy.transformRec(fn);
+      if (!newParentTy)
         return Type();
-
-      if (substParentTy.getPointer() != parentTy.getPointer())
+      if (newParentTy.getPointer() != parentTy.getPointer())
         anyChanged = true;
     }
 
-    for (const auto arg : bound->getDirectGenericArgs()) {
-      Type substArg = arg.transformRec(fn);
-      if (!substArg)
+    // Attempt to transform only the direct generic arguments.
```

Comment:
Note that subst() should now special-case BoundGenericType, by calling subst() on the stored substitution map, instead of doing what transform() does here.